### PR TITLE
Adding modifyloadbalancer api support in Windows KubeProxy for update loadbalancer replacing usage of delete and create loadbalancer api.

### PR DESCRIPTION
**Pull request by princepereira**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
We're implementing container networking changes aimed at improving performance in Server 2025. To accommodate these modifications, we need to make minor adjustments to Kubeproxy, particularly regarding certain HCN API calls for updating load balancers and deleting remote endpoints. Without these updates, certain functionalities, such as proxy terminating endpoints, would be broken on newer versions of Windows OS.

To elaborate, this PR encompasses two main changes:

Previously, updating the load balancer policy involved deleting and recreating the load balancer. Now, we're leveraging a pre-existing update load balancer policy API, which was not previously implemented in our control plane, to simplify the update process.

Additionally, this PR aims to streamline the management of remote endpoint lifecycles. Previously, unready pods underwent unnecessary multiple deletion and recreation phases. With these adjustments, each IP will now undergo a single create and delete cycle for remote endpoints.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?
NONE

```release-note

- Windows Kubeproxy will use the update load balancer API for load balancer updates, instead of the previous delete and create APIs.
- Deletion of remote endpoints will be triggered only for terminated endpoints (those present in the old endpoints map but not in the new endpoints map), whereas previously it was also done for terminating endpoints.

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/proxy/winkernel/proxier.go [hunk 1]
```diff
@@ -348,14 +362,30 @@ func (proxier *Proxier) endpointsMapChange(oldEndpointsMap, newEndpointsMap prox
...
+		}
 		svcPortMap[svcPortName] = true
 		proxier.onEndpointsMapChange(&svcPortName, false)
 	}
 
 	for svcPortName, eps := range newEndpointsMap {
 		logFormattedEndpoints("endpointsMapChange newEndpointsMap", logLevel, svcPortName, eps)
+		if proxier.winProxyEbpfMode {
+			for _, ep := range eps {
+				if !ep.IsLocal() && proxier.remoteEPsToCleanup[ep.IP()] {
```
Comment (sourad) [inline pkg/proxy/winkernel/proxier.go:384]: Probably isLocal check is not needed here. ep in remoteEPsToCleanup shouldn't be local per check in line 371. And if at all ep is local and in remoteEPsToCleanup, we would ideally like to remove it from remoteEPsToCleanup?

File: pkg/proxy/winkernel/proxier.go [hunk 2]
```diff
@@ -348,14 +362,30 @@ func (proxier *Proxier) endpointsMapChange(oldEndpointsMap, newEndpointsMap prox
 	// This will optimize remote endpoint and loadbalancer deletion based on the annotation
```
Comment (sourad) [inline pkg/proxy/winkernel/proxier.go:362]: As discussed, please add a comment on when endpointsMapChange gets called.

**Comment by aojea** [comment]
/hold

is there a KEP? I can not find it, also windows proxy is not well maintained so we should understand the implications of adding more functionality if there is no commitment from people to maintain it

**Comment by aojea** [comment]
/hold

I think we need to have a conversation about the windows kube-proxy in tree and define a better plan and ownership, it is kind of an opaque component right now and when there are bugs we always have doubts to whom we should assign them

/cc @shaneutt @thockin @danwinship

**Comment by shaneutt** [comment]
Yes sounds good. I added it to our SIG meeting agenda for the next sync.

**Comment by princepereira** [comment]
> /hold
> 
> I think we need to have a conversation about the windows kube-proxy in tree and define a better plan and ownership, it is kind of an opaque component right now and when there are bugs we always have doubts to whom we should assign them
> 
> /cc @shaneutt @thockin @danwinship

@aojea , @shaneutt ,

Sravanth ( @sbangari ), David ( @daschott ) and I ( @princepereira ) are the owners. Please feel free to assign any Windows Kube Proxy bugs to us.

**Comment by princepereira** [comment]
> /hold
> 
> is there a KEP? I can not find it, also windows proxy is not well maintained so we should understand the implications of adding more functionality if there is no commitment from people to maintain it

@aojea ,

Apologies for the delayed response. 

To provide some context, we're implementing container networking changes aimed at improving performance in Server 2025. To accommodate these modifications, we need to make minor adjustments to Kubeproxy, particularly regarding certain HCN API calls for updating load balancers and deleting remote endpoints. **Without these updates, certain functionalities, such as proxy terminating endpoints, would be broken on newer versions of Windows OS.** It's important to note that these changes are not intended to introduce new functionalities but rather to ensure compatibility with the latest Windows OS versions. 

To elaborate, this PR encompasses two main changes: 

Previously, updating the load balancer policy involved deleting and recreating the load balancer. Now, we're leveraging a pre-existing modify load balancer policy API, which was not previously implemented in our control plane, to simplify the update process. 

Additionally, this PR aims to streamline the management of remote endpoint lifecycles. Previously, unready pods underwent unnecessary multiple deletion and recreation phases. With these adjustments, each IP will now undergo a single create and delete cycle for remote endpoints. 

The feature gate in the initial PR may have been a bit misleading. **The initial changes were in a draft state and were solely intended for internal proof of concept and validation**. There was no intention of introducing feature gates or KEP in the final PR, as no new features are being introduced. Please let me know if this seems alright to you.

**Comment by aojea** [comment]
> The feature gate in the initial PR may have been a bit misleading. **The initial changes were in a draft state and were solely intended for internal proof of concept and validation**. There was no intention of introducing feature gates or KEP in the final PR, as no new features are being introduced. Please let me know if this seems alright to you.

sounds good to me, but please keep in mind that this winproxy is within the sig network umbrella and it can not operate in an island mode, sending an email to the mailing list or presenting this work in the sig network meeting can be good ways of introducing this ideas and work, otherwise for reviewers PRs like this, are hard to understand without the context you are providing us now

/hold cancel

**Comment by princepereira** [comment]
> > The feature gate in the initial PR may have been a bit misleading. **The initial changes were in a draft state and were solely intended for internal proof of concept and validation**. There was no intention of introducing feature gates or KEP in the final PR, as no new features are being introduced. Please let me know if this seems alright to you.
> 
> sounds good to me, but please keep in mind that this winproxy is within the sig network umbrella and it can not operate in an island mode, sending an email to the mailing list or presenting this work in the sig network meeting can be good ways of introducing this ideas and work, otherwise for reviewers PRs like this, are hard to understand without the context you are providing us now
> 
> /hold cancel

@aojea ,

The design had not been finalized, and the code was still in the proof-of-concept stage. Therefore, we kept it in draft form and have not published it yet. We assumed that emails were only required for the published PRs. We intend to publish and share the details via email once the feasibility and design are confirmed. Thank you for understanding.

File: pkg/proxy/winkernel/proxier.go [hunk 3]
```diff
@@ -1616,23 +1642,147 @@ func (proxier *Proxier) deleteExistingLoadBalancer(hns HostNetworkService, winPr
...
+	return true
 }
 
-func (proxier *Proxier) deleteLoadBalancer(hns HostNetworkService, lbHnsID *string) bool {
-	klog.V(3).InfoS("Hns LoadBalancer delete triggered for loadBalancer resources", "lbHnsID", *lbHnsID)
-	if err := hns.deleteLoadBalancer(*lbHnsID); err != nil {
-		// This will be cleanup by cleanupStaleLoadbalancer fnction.
-		proxier.mapStaleLoadbalancers[*lbHnsID] = true
+// configureLoadbalancer takes care of deleting and creating lb policy in old hns architecture and update policy in
+// new architecture. If there are no backend endpoints, then code will only delete existing loadbalancer.
```
[the remark this answers is not in the document]
Comment (princepereira) [reply pkg/proxy/winkernel/proxier.go]: These changes are no more part of pr.

File: pkg/proxy/winkernel/proxier.go (hunk 1 above)
Comment (princepereira) [reply pkg/proxy/winkernel/proxier.go]: These changes are no more part of pr.

File: pkg/proxy/winkernel/proxier.go [hunk 4]
```diff
@@ -1616,23 +1642,147 @@ func (proxier *Proxier) deleteExistingLoadBalancer(hns HostNetworkService, winPr
...
+			lbFlags,
+			protocol,
+			targetPort,
+			extPort,
+			queriedLoadBalancers,
+		)
+		return
+	}
+
+	klog.V(3).InfoS("Create and delete loadbalancer called.", "lbHnsID", *lbHnsID, "sourceVip", sourceVip, "vip", vip, "port", targetPort, "lbType", lbType)
```
[the remark this answers is not in the document]
Comment (princepereira) [reply pkg/proxy/winkernel/proxier.go]: Changes are not part of new pr.

File: pkg/proxy/winkernel/proxier.go (hunk 2 above)
Comment (princepereira) [reply pkg/proxy/winkernel/proxier.go:366]: Addressed.

**Review by sbangari — APPROVED** [review]

**Pull request merged**